docs/source/developer/contributing.rst

```diff
@@ -1,19 +1,627 @@
+function to get a ``Promise`` for a ``requestAnimationFrame``. We
+sometimes have to set a sentinel value inside a ``Promise`` and then
+check that the sentinel was set if we need a promise to run without
+blocking.
+
+Performance Testing
+~~~~~~~~~~~~~~~~~~~
+
+If you are making a change that might affect how long it takes to load
+JupyterLab in the browser, we recommend doing some performance testing
+using `Lighthouse <https://github.com/GoogleChrome/lighthouse>`__. It
+let's you easily compute a number of metrics, like page load time, for
+the site.
+
+To use it, first build JupyterLab in dev mode:
+
+.. code:: bash
+
+   jlpm run build:dev
+
+Then, start JupyterLab using the dev build:
+
+.. code:: bash
+
+   jupyter lab --dev --NotebookApp.token=''  --no-browser
+
+Now run Lighthouse against this local server and show the results:
+
+.. code:: bash
+
+   jlpm run lighthouse --view
+
+.. image:: /images/lighthouse.png
+
+Using throttling
+''''''''''''''''
+
+Lighthouse recommends using the system level
+`comcast <https://github.com/tylertreat/comcast>`__ tool to throttle
+your network connection and emulate different scenarios. To use it,
+first install that tool using ``go``:
+
+.. code:: bash
+
+   go get github.com/tylertreat/comcast
+
+Then, before you run Lighthouse, enable the throttling (this requires
+sudo):
+
+.. code:: bash
+
+   run lighthouse:throttling:start
+
+This enables the "WIFI (good)" preset of comcast, which should emulate
+loading JupyterLab over a local network.
+
+Then run the lighthouse tests:
+
+.. code:: bash
+
+   jlpm run lighthouse [...]
+
+Then disable the throttling after you are done:
+
+.. code:: bash
+
+   jlpm run lighthouse:throttling:stop
+
+Comparing results
+'''''''''''''''''
+
+Performance results are usually only useful in comparison to other
+results. For that reason, we have included a comparison script that can
+take two lighthouse results and show the changes between them.
+
+Let's say we want to compare the results of the production build of
+JupyterLab with the normal build. The production build minifies all the
+JavaScript, so should load a bit faster.
+
+First, we build JupyterLab normally, start it up, profile it and save
+the results:
+
+.. code:: bash
+
+   jlpm build:dev
+   jupyter lab --dev --NotebookApp.token='' --no-browser
+
+   # in new window
+   jlpm run lighthouse --output json --output-path normal.json
+
+Then rebuild with the production build and retest:
+
+.. code:: bash
+
+   jlpm run build:dev:prod
+   jupyter lab --dev --NotebookApp.token='' --no-browser
+
+   # in new window
+   jlpm run lighthouse --output json --output-path prod.json
+
+Now we can use compare the two outputs:
+
+.. code:: bash
+
+   jlpm run lighthouse:compare normal.json prod.json
+
+This gives us a report of the relative differences between the audits in
+the two reports:
+
+.. admonition:: Resulting Output
+
+   | normal.json -> prod.json
```

Comment:
this line and the following report seem to have lost some of their previous formatting (though it's a bit hard to tell from github's "rich" preview)